### PR TITLE
FXIOS-2981 - Business Logic and UI for Search Term Groups [iPhone and iPad] 

### PR DESCRIPTION
**Pull request by nbhasin2**
Added Search Term Groups in Tabs Tray which has the following sections and is only applicable for Normal / Active tabs and not for private mode.

Note: This PR also turns on search term metadata collection (FXIOS-3063)

## a) Search Term Groups
This section contains grouped tabs formed by the search term entered by the user

iPhone: Number of items shown on screen in a group are 2 unless we go into landscape mode. 
iPad: Number of items shown on screen in a group are 4

Min number of items that can make a group for both iPhone and iPad are 2. If only a single Tab is part of search metadata group then we don't do anything with it and simply make it a part of Regular Tab

Usage:
-- User opens a new tab and searches for a _Search Term 1_
-- This _Search Term 1_ gets recorded as tab history metadata 
-- User opens another tab and searches for a _Search Term 1_ (same search) but now taps on a item on search page. This is a search follow-on scenario
-- We combine both tabs into **Search Term 1 Group** in Tab Tray
-- Search Groups are collection of tabs when user searches for a particular item from *Awesome Bar* 

## b) Regular Tabs

These are normal tabs that are not part of a group and user simply opens them by entering URL or long pressing on a link. 

## c) Inactive Tabs

Inactive tabs (if enabled) should put a tab to inactive mode after 4 days. The number of days after which a tab gets inactive might change in future. 

**Whats missing:** 
Telemetry and UI polish

### CODE REVIEW

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 1]
```diff
@@ -1047,6 +1047,11 @@ class BrowserViewController: UIViewController {
             // the same origin as the current URL. Otherwise, do nothing and wait for
             // didCommitNavigation to confirm the page load.
             if tab.url?.origin == webView.url?.origin {
+                
+//                if tab == tabManager.selectedTab, tab.url?.absoluteString != webView.url?.absoluteString {
+//                    tab.updateTimerAndObserving(state: .tabNavigatedToDifferentUrl)
+//                }
```
Comment (electricRGB) [inline Client/Frontend/Browser/BrowserViewController.swift:1053]: Should we remove this, seeing as you moved it down?

File: Client/Frontend/Browser/SearchViewController.swift [hunk 2]
```diff
@@ -465,19 +465,23 @@ class SearchViewController: SiteTableViewController, KeyboardHelperDelegate, Loa
             if let url = engine.searchURLForQuery(suggestion) {
                 Telemetry.default.recordSearch(location: .suggestion, searchEngine: engine.engineID ?? "other")
                 GleanMetrics.Search.counts["\(engine.engineID ?? "custom").\(SearchesMeasurement.SearchLocation.suggestion.rawValue)"].add()
-                
-                searchDelegate?.searchViewController(self, didSelectURL: url)
+                let key: HistoryMetadataKey = HistoryMetadataKey(url: url.absoluteString, searchTerm: suggestion, referrerUrl: "")
+//                let observation: HistoryMetadataObservation = HistoryMetadataObservation(titleObservation: "sometitle", viewTimeObservation: 2, documentTypeObservation: .regular)
+//                let metaDataKey = HistoryMetadataKey(url: "", searchTerm: "", referrerUrl: "")
+//                let observation = HistoryMetadataObservation()
+//                profile.places.recordHistoryMetaData(key: key, observation: observation)
```
Comment (electricRGB) [inline Client/Frontend/Browser/SearchViewController.swift:472]: Not sure what's happening here

File: Client/Frontend/Browser/Tab.swift [hunk 3]
```diff
@@ -148,6 +168,7 @@ class Tab: NSObject {
         didSet {
             if let _url = url, let internalUrl = InternalURL(_url), internalUrl.isAuthorized {
                 url = URL(string: internalUrl.stripAuthorization)
+                print("URLLLLL - SET \(url)")
```
Comment (electricRGB) [inline Client/Frontend/Browser/Tab.swift:171]: Bad print statement, bad! :P

File: Client/Frontend/Browser/TabDisplayManager.swift [hunk 4]
```diff
@@ -283,13 +351,15 @@ class TabDisplayManager: NSObject, FeatureFlagsProtocol {
 
 extension TabDisplayManager: UICollectionViewDataSource {
     @objc func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
-        if !shouldEnableInactiveTabs { return dataStore.count }
+//        if !shouldEnableInactiveTabs { return dataStore.count }
```
Comment (electricRGB) [inline Client/Frontend/Browser/TabDisplayManager.swift:354]: Do we still need this?

File: Client/Frontend/Browser/TabDisplayManager.swift [hunk 5]
```diff
@@ -161,21 +175,68 @@ class TabDisplayManager: NSObject, FeatureFlagsProtocol {
...
+//        }
+        
+        
+//        isInactiveViewExpanded = inactiveViewModel.inactiveTabs.count > 0
+//        let recentlyClosedTabs = inactiveViewModel.recentlyClosedTabs
+//        if recentlyClosedTabs.count > 0 {
+//            tabManager.removeTabs(recentlyClosedTabs, shouldNotify: true)
+//            tabManager.selectTab(selectedTab)
+//        }
+//        return inactiveViewModel.activeTabs
```
Comment (electricRGB) [inline Client/Frontend/Browser/TabDisplayManager.swift:239]: Still needed?

File: Client/Frontend/Browser/TabDisplayManager.swift [hunk 6]
```diff
@@ -321,9 +399,9 @@ extension TabDisplayManager: UICollectionViewDataSource {
     }
     
     @objc func numberOfSections(in collectionView: UICollectionView) -> Int {
-        if !shouldEnableInactiveTabs { return 1 }
+//        if !shouldEnableInactiveTabs { return 1 }
```
Comment (electricRGB) [inline Client/Frontend/Browser/TabDisplayManager.swift:402]: Neded?

File: Client/Frontend/Browser/TabDisplayManager.swift [hunk 7]
```diff
@@ -85,17 +88,26 @@ class TabDisplayManager: NSObject, FeatureFlagsProtocol {
     private let tabReuseIdentifer: String
     var profile: Profile
     private var inactiveNimbusExperimentStatus: Bool = false
+    var shouldEnableGroupedTabs: Bool = true
+    
+//    {
+//        guard featureFlags.isFeatureActive(.groupedTabs) else { return false }
+//        
+//        return inactiveNimbusExperimentStatus ? inactiveNimbusExperimentStatus : profile.prefs.boolForKey(PrefsKeys.KeyEnableInactiveTabs) ?? false
+//    }
```
[the remark this answers is not in the document]
Comment (nbhasin2) [reply Client/Frontend/Browser/TabDisplayManager.swift:97]: Yeah, a lot of comments are being removed in my next commit. Had to do this so I can fully test it out :)

File: Client/Frontend/Browser/Tabs/GroupedTabCell.swift [hunk 8]
```diff
@@ -268,7 +274,10 @@ class GroupedTabContainerCell: UITableViewCell, UICollectionViewDelegateFlowLayo
 
     @objc func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
         let cellWidth = floor((collectionView.bounds.width - GridTabTrayControllerUX.Margin * CGFloat(numberOfColumns + 1)) / CGFloat(numberOfColumns))
-        let padding = 10
+        let isIpad = UIDevice.current.userInterfaceIdiom == .pad
+        let orientation = UIDevice.current.orientation
+        let isPortrait = (orientation == .portrait || orientation == .portraitUpsideDown)
```
Comment (electricRGB) [inline Client/Frontend/Browser/Tabs/GroupedTabCell.swift:279]: There's an easier way to do this, if you want. I added a `.isLandscape` property on UIWindow. :) No need to change it; just letting you know. :)

Comment (nbhasin2) [reply Client/Frontend/Browser/Tabs/GroupedTabCell.swift]: @electricRGB Thanks Roux, didn't know that so I updated it :)

**Review by electricRGB — APPROVED** [review]

**Comment by nbhasin2** [comment]
Note: cleanup commit message during squash and merge

**Comment by Psoffie009** [comment]
> Added Search Term Groups in Tabs Tray which has the following sections and is only applicable for Normal / Active tabs and not for private mode.
> 
> Note: This PR also turns on search term metadata collection (FXIOS-3063)
> 
> ## a) Search Term Groups
> This section contains grouped tabs formed by the search term entered by the user
> 
> iPhone: Number of items shown on screen in a group are 2 unless we go into landscape mode.
> iPad: Number of items shown on screen in a group are 4
> 
> Min number of items that can make a group for both iPhone and iPad are 2. If only a single Tab is part of search metadata group then we don't do anything with it and simply make it a part of Regular Tab
> 
> Usage:
> -- User opens a new tab and searches for a _Search Term 1_
> -- This _Search Term 1_ gets recorded as tab history metadata
> -- User opens another tab and searches for a _Search Term 1_ (same search) but now taps on a item on search page. This is a search follow-on scenario
> -- We combine both tabs into **Search Term 1 Group** in Tab Tray
> -- Search Groups are collection of tabs when user searches for a particular item from _Awesome Bar_
> 
> ## b) Regular Tabs
> These are normal tabs that are not part of a group and user simply opens them by entering URL or long pressing on a link.
> 
> ## c) Inactive Tabs
> Inactive tabs (if enabled) should put a tab to inactive mode after 4 days. The number of days after which a tab gets inactive might change in future.
> 
> **Whats missing:**
> Telemetry and UI polish

File: Client/Frontend/Browser/GridTabViewController.swift [hunk 9]
```diff
@@ -304,7 +306,7 @@ extension GridTabViewController {
     }
 
     func closeTabsForCurrentTray() {
-        let tabs = self.tabDisplayManager.dataStore.compactMap { $0 }
+        let tabs = self.tabDisplayManager.tabsInAllGroups ?? [Tab]() + self.tabDisplayManager.filteredTabs
```
Comment (nishant2718) [inline Client/Frontend/Browser/GridTabViewController.swift:309]: I believe `[Tab]()` is preventing us from closing non-private tabs.

**Pull request merged**